Support certs with separate Extended Key Usage

lavacat opened the pull request:
Problem: Right now certs that specify extended key usage must set
both clientAuth and serverAuth.

Solution: Add settings to path additional server cert and use it when
creating ssl server

*Issue #474 *

*Description of changes:*


By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

vrozov commented:
@lavacat Please check why CI build failed

debjanibnrj commented on src/main/java/com/amazon/opendistroforelasticsearch/security/ssl/OpenDistroSecuritySSLPlugin.java:
```diff
@@ -329,6 +329,12 @@ public Object run() {
         settings.add(Setting.simpleString(SSLConfigConstants.OPENDISTRO_SECURITY_SSL_TRANSPORT_PEMKEY_PASSWORD, Property.NodeScope, Property.Filtered));
         settings.add(Setting.simpleString(SSLConfigConstants.OPENDISTRO_SECURITY_SSL_TRANSPORT_PEMTRUSTEDCAS_FILEPATH, Property.NodeScope, Property.Filtered));
 
+        settings.add(Setting.boolSetting(SSLConfigConstants.OPENDISTRO_SECURITY_SSL_TRANSPORT_MIXED_KEY_USAGE_ENABLED, SSLConfigConstants.OPENDISTRO_SECURITY_SSL_TRANSPORT_MIXED_KEY_USAGE_ENABLED_DEFAULT, Property.NodeScope, Property.Filtered));
```
it may be good to disable these settings if mixedKeyUsage is false. Eg - 

```suggestion
    if (!isMixedKeyUsageAllowed(settings)) {
          settings.add(Setting.simpleString(SSLConfigConstants.OPENDISTRO_SECURITY_SSL_TRANSPORT_SERVER_PEMCERT_FILEPATH, Property.NodeScope, Property.Filtered));
          settings.add(Setting.simpleString(SSLConfigConstants.OPENDISTRO_SECURITY_SSL_TRANSPORT_SERVER_PEMKEY_FILEPATH, Property.NodeScope, Property.Filtered));
          settings.add(Setting.simpleString(SSLConfigConstants.OPENDISTRO_SECURITY_SSL_TRANSPORT_SERVER_PEMKEY_PASSWORD, Property.NodeScope, Property.Filtered));
          settings.add(Setting.simpleString(SSLConfigConstants.OPENDISTRO_SECURITY_SSL_TRANSPORT_SERVER_PEMTRUSTEDCAS_FILEPATH, Property.NodeScope, Property.Filtered));
        }
```

You can create a helper function like - 
```
    private static boolean isMixedKeyUsageAllowed(final Settings settings) {
        return settings.getAsBoolean(SSLConfigConstants. OPENDISTRO_SECURITY_SSL_TRANSPORT_MIXED_KEY_USAGE_ENABLED, false);
    }
```

lavacat commented:
@debjanibnrj Thanks for comments. I'll resume working on this. 
- will add JKS cert support
- mixedKeyUsageEnabled means both clientAuth and serverAuth, if it's confusing, happy to use some other name
- agree with having .server. and .client. and defaulting to old way. Will refactor

nicot commented on src/test/resources/ssl/extended_key_usage/README.txt:
```diff
@@ -0,0 +1,22 @@
+# Commands to generate certs in this folder
```
Thanks for adding this!

nicot reviewed (dismissed):
Looks great to me! Thanks for making these changes.

debjanibnrj commented on src/main/java/com/amazon/opendistroforelasticsearch/security/ssl/OpenDistroSecuritySSLPlugin.java:
```diff
@@ -346,6 +350,24 @@ public Object run() {
+            settings.add(Setting.simpleString(SSLConfigConstants.OPENDISTRO_SECURITY_SSL_TRANSPORT_SERVER_KEYSTORE_KEYPASSWORD, Property.NodeScope, Property.Filtered));
+            settings.add(Setting.simpleString(SSLConfigConstants.OPENDISTRO_SECURITY_SSL_TRANSPORT_CLIENT_KEYSTORE_KEYPASSWORD, Property.NodeScope, Property.Filtered));
+
+            settings.add(Setting.simpleString(SSLConfigConstants.OPENDISTRO_SECURITY_SSL_TRANSPORT_SERVER_PEMCERT_FILEPATH, Property.NodeScope, Property.Filtered));
```
Just for my knowledge, is it ever possible for Http Certs to use different client and server certs?

lavacat replied:
If I understand your question correctly, the answer is no. 
We only buildSSLServerContext in initHttpSSLConfig, so if users want to use different client/server certs they can provide their server cert using current config options.

debjanibnrj commented on src/main/java/com/amazon/opendistroforelasticsearch/security/ssl/util/CertFromKeystore.java:
```diff
@@ -0,0 +1,113 @@
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.security.*;
```
avoid wildcard imports

debjanibnrj commented on src/main/java/com/amazon/opendistroforelasticsearch/security/ssl/util/CertFromKeystore.java:
```diff
@@ -0,0 +1,113 @@
+
+    public CertFromKeystore(KeystoreProps keystoreProps, String keystoreAlias, String keyPassword) throws CertificateException, NoSuchAlgorithmException, KeyStoreException, IOException, UnrecoverableKeyException {
+        this.keystoreProps = keystoreProps;
+        final KeyStore ks = keystoreProps.loadKeystore();;
```
nit: double ;

lavacat commented on src/main/java/com/amazon/opendistroforelasticsearch/security/ssl/DefaultOpenDistroSecurityKeyStore.java:
```diff
@@ -427,30 +446,12 @@ public void initHttpSSLConfig() {
-                X509Certificate[] trustedHTTPCertificates = null;
+                CertFromKeystore certFromKeystore = new CertFromKeystore(keystoreProps, keystoreAlias, keyPassword);
 
+                CertFromTruststore certFromTruststore = CertFromTruststore.Empty();
```
this is just a way to say that all properties of certFromTruststore are null. 
I could have initialed certFromTruststore to null, but then I'd need another if check when buildSSLServerContext below. Instead I just call certFromTruststore.getServerTrustedCerts() that can return null.

Previous code relies on the fact that trustedHTTPCertificates can be null.

lavacat commented:
@nicot @debjanibnrj 
- added keystore, truststore files and integration test in SSLTest.java
- added unit tests for all new CertFrom* classes
- cleaned up imports

@debjanibnrj based on your question "Why add || extendedKeyUsageEnabled here?"
did a small refactor and added cleaner useKeyStore and useRawFiles variables

Nothing else changed

debjanibnrj reviewed (dismissed):
Thanks for your contribution. Added a minor comment but overall looks great!

debjanibnrj commented:
@hardik-k-shah can you have a look as well

lavacat commented:
> Thanks for your contribution. Added a minor comment but overall looks great!

@debjanibnrj FYI, don't see the new comments. I think I've addressed all older ones.

debjanibnrj commented on src/main/java/com/amazon/opendistroforelasticsearch/security/ssl/DefaultOpenDistroSecurityKeyStore.java:
```diff
@@ -293,107 +294,135 @@ public void initTransportSSLConfig() {
                 logExplanation(e);
                 throw new ElasticsearchSecurityException(
-                    "Error while initializing transport SSL layer: " + e.toString(), e);
+                        "Error while initializing transport SSL layer from PEM: " + e.toString(), e);
```
nit: these are not PEM certificates.

lavacat replied:
thanks for noticing, reverted back

hardik-k-shah commented on src/main/java/com/amazon/opendistroforelasticsearch/security/ssl/DefaultOpenDistroSecurityKeyStore.java:
```diff
@@ -293,107 +294,135 @@ public void initTransportSSLConfig() {
+                            null);
+                    final String keyPassword = settings.get(SSLConfigConstants.OPENDISTRO_SECURITY_SSL_TRANSPORT_KEYSTORE_KEYPASSWORD,
+                            keystorePassword);
+
```
should we also check truststoreAlias and keystoreAlias for non-null and throw exception if anyone of it was set to null?

lavacat replied:
If the alias isn't provided SSLCertificateHelper.exportServerCertChain will take the first one. I've made it required for extended key usage because server/client alias should be different at least for keystore.
Made a comment and added a test

hardik-k-shah commented on src/test/java/com/amazon/opendistroforelasticsearch/security/ssl/util/CertFromFileTests.java:
```diff
@@ -0,0 +1,57 @@
+
+    Assert.assertNull(cert.getClientTrustedCas());
+  }
+
```
Can you also add one negative test to verify if extended key is enabled but either server/client cert file or both are missing ?

hardik-k-shah reviewed (commented):
Thanks @lavacat and @nicot for your contribution. 
Overall it looks good. Few minor comments.

lavacat commented:
@debjanibnrj @hardik-k-shah thanks for comments and noticing those mistakes, made all changes as you've suggested.
